### PR TITLE
Switch all x86 and RISC-V matmul codegen to use DoubleTilingExpert.

### PR DESCRIPTION
**Pull request by hanhanW**
The RISC-V targets were considered as ARM configuration. This PR makes
it go through Sandbox based approach.

The commit also moves quantized matmul to use DoubleTilingExpert. They
used old pipeline because of long compilation time. Many instructions
were generated during the lowering. This is addressed in https://github.com/llvm/llvm-project/commit/1538bd518cd236f4321695e9c5f0dd24601db366

It is a step toward https://github.com/google/iree/issues/8431

### CODE REVIEW

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (MaheshRavishankar) [inline iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:475]: If I understand it correctly, the `setX86TileFuseAndVectorizeRootConfig` is now dead. We can delete it (and also the corresponding pass pipeline?)

**Comment by MaheshRavishankar** [comment]
Just curious, does it help the compilation time at all.

Aside : Really great that this is all now consolidated. Lets see what the perf looks like.

**Comment by hanhanW** [comment]
> Just curious, does it help the compilation time at all.

Yes, this helps the compilation time a lot.  The main benefits are come from better tiling sizes. I measured that it takes 6x sec when going through TileFuseAndVectorize pipeline, and 4x sec when going through DoubleTilingExpert.

I think this is because we have much more cleanup in new pipelines. (It's been a long time that we dont touch the old pipeline. I'm not surprised if we miss some patterns today.)

File: iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (hanhanW) [reply iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp:475]: Good catch!

**Review by dcaballe — APPROVED** [review]
Thanks, LGTM! Please, wait for Mahesh's approval.

> The main benefits are come from better tiling sizes.

Better is which regard? Are they smaller and, therefore, less unrolling happens? If the compile time is still unexpectedly high I would suggest taking a look at what LLVM is doing. There are a couple of unrolling passes that could be kicking in regardless of what we do at the MLIR level. We found similar compile time issues in XLA due to that.

**Comment by hanhanW** [comment]
> Better is which regard? Are they smaller and, therefore, less unrolling happens? If the compile time is still unexpectedly high I would suggest taking a look at what LLVM is doing. There are a couple of unrolling passes that could be kicking in regardless of what we do at the MLIR level. We found similar compile time issues in XLA due to that.

Yes, smaller tiling sizes are selected, so less unrolling happens. They used to be `{1, ..., 5v, v, 16v}` where v is the native vector size. Also, the new pipeline has more cleanups. I think the compilation time is fine now, we already cut it from dozens of minutes to dozens of seconds. We can do another round optimization later. :)

Re Mahesh:

For RV64 it takes 7 seconds now (versus 20 seconds before).

image

I could not find numbers for RV32 targets in CI. I ran benchmarks locally.

Without RVV features, we cut the compilation time from `>5 mins` to 20 secs. With RVV features, we cut the compilation time from `>25 mins` to 45 secs.

**Comment by hanhanW** [comment]
> Just curious, does it help the compilation time at all.
> 
> Aside : Really great that this is all now consolidated. Lets see what the perf looks like.

ah, just realized that you meant latency perf, not perf of compilation time. We won't see it from BuildKite benchmarks because RISC-V targets are not included. I think we don't track RISC-V perf at all.

**Comment by MaheshRavishankar** [comment]
> > Just curious, does it help the compilation time at all.
> > Aside : Really great that this is all now consolidated. Lets see what the perf looks like.
> 
> ah, just realized that you meant latency perf, not perf of compilation time. We won't see it from BuildKite benchmarks because RISC-V targets are not included. I think we don't track RISC-V perf at all.

Ah I thought even the ARM path was changed. Ok, makes sense now.

**Review by MaheshRavishankar — APPROVED** [review]

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**
